### PR TITLE
Transaction logs for data source changes

### PR DESCRIPTION
**Pull request by SmittieC**
[Ticket](https://dimagi-dev.atlassian.net/browse/SC-3066)

## Technical Summary

- Added a new table (`DataSourceRowTransactionLog`) in userreports used to keep track of any changes that happens in a datasource table.
- Added a new Repeater Type - `DataSourceRepeater`. This repeater is used to forward any data source changes to a client.
- Added a new data source chaned signal that gets called when upserting or deleting something from a datasource table.
- When a data source is changed, we check if the data source is "subscibed" to by checking if a Repeater with this data source id exists. If True, we create a `DataSourceRowTransactionLog` entry and a repeat record.

## Feature Flag

[SUPERSET_ANALYTICS](https://www.commcarehq.org/hq/flags/edit/superset-analytics/)

### Safety story

- Local testing
- I expanded some pillow tests (the ones that tests data sources) to also test that `DataSourceRowTransactionLog` entries are being made.
- Staging testing confirms that the repeater sends the payload. If you're a curious beaver, have a look [here](https://staging.commcarehq.org/a/develop/settings/project/couch_repeat_record_report/?repeater=4bb34504-c0e5-42a3-986f-eac23d48549f&record_state=&payload_id=). (The failing ones are from earlier testing)

### QA Plan

This cannot be QA'd in a meaningful manner yet.

### Migrations

- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [ ] This PR can be reverted after deploy with no further considerations

We might want to remove
- the new `DataSourceRepeater` table
- the new `DataSourceRowTransactionLog` table

Not doing that should not break anything though

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Review by snopoke — COMMENTED** [review]
initial review looks good

File: corehq/apps/userreports/tests/test_pillow.py
Comment (SmittieC) [inline corehq/apps/userreports/tests/test_pillow.py:386]: I'm doing this, since `BaseRepeaterTest` set's up a domain for us, ready to use/handle repeaters

**Comment by SmittieC** [comment]
Hmm, the failing test is passing locally :thinking:

**Review by zandre-eng — COMMENTED** [review]
Overall looks good, just a few minor comments.

File: corehq/motech/repeaters/models.py
Comment (SmittieC) [reply corehq/motech/repeaters/models.py:871]: Yeah, but there are others that does not result in circular imports. ima fix that

File: corehq/motech/repeaters/signals.py
Comment (SmittieC) [reply corehq/motech/repeaters/signals.py:109]: yeah

File: corehq/apps/userreports/tests/test_pillow.py
Comment (SmittieC) [reply corehq/apps/userreports/tests/test_pillow.py:457]: You're right, we don't, thanks!

File: corehq/apps/userreports/tests/test_pillow.py
Comment (SmittieC) [reply corehq/apps/userreports/tests/test_pillow.py:556]: Nope, just forgot that `assertEqual` exists at that time!

File: corehq/motech/repeaters/models.py
Comment (SmittieC) [reply corehq/motech/repeaters/models.py]: https://github.com/dimagi/commcare-hq/pull/33781/commits/63a9f23802a6c6d9571e628684274c1bec201c97

File: corehq/apps/userreports/tests/test_pillow.py
Comment (SmittieC) [reply corehq/apps/userreports/tests/test_pillow.py]: https://github.com/dimagi/commcare-hq/pull/33781/commits/63a9f23802a6c6d9571e628684274c1bec201c97

File: corehq/motech/repeaters/models.py
Comment (SmittieC) [reply corehq/motech/repeaters/models.py:870]: https://github.com/dimagi/commcare-hq/pull/33781/commits/63a9f23802a6c6d9571e628684274c1bec201c97

File: corehq/motech/repeaters/models.py
Comment (zandre-eng) [inline corehq/motech/repeaters/models.py:418]: It seems you've managed to resolve the issue with the failing test, so just noting that this debug log should be removed.

**Review by zandre-eng — COMMENTED** [review]
Looks good. Just one comment that needs addressing before approval.

File: corehq/motech/repeaters/models.py
Comment (SmittieC) [reply corehq/motech/repeaters/models.py]: Ah thanks!

**Review by zandre-eng — APPROVED** [review]

File: corehq/apps/userreports/tests/test_pillow.py
Comment (mkangia) [inline corehq/apps/userreports/tests/test_pillow.py:408]: why is this needed?
wont any new records created during test be removed at the end of the test?

File: corehq/apps/userreports/util.py
Comment (mkangia) [inline corehq/apps/userreports/util.py:365]: Can we move this if block to where the function is getting called since this is at the top of the function definition?

**Review by mkangia — COMMENTED** [review]
Took a quick pass.
I am happy to review this properly if needed. I don't have context on Data Source Repeaters so I would need to learn that a bit before being able to review this properly.

File: corehq/apps/userreports/util.py
Comment (SmittieC) [reply corehq/apps/userreports/util.py:365]: The only reason why I put it in here is to reduce code duplication, since this function is called from two places. If you feel strongly about this, I'm happy to move it..

File: corehq/apps/userreports/tests/test_pillow.py
Comment (kaapstorm) [inline corehq/apps/userreports/tests/test_pillow.py:390]: Instead of changing the domain name of `domain_obj` after it has been created, just override its name on the class, and it will be created by `super().setUpClass()` with the correct name.

```suggestion
class IndicatorPillowTest(BaseRepeaterTest):
    domain = 'user-reports'  # Domain name from sample data source

    @classmethod
    def setUpClass(cls):
        super(IndicatorPillowTest, cls).setUpClass()
        cls.config = get_sample_data_source()
        cls.config.save()
```

File: corehq/apps/userreports/tests/test_pillow.py
Comment (kaapstorm) [inline corehq/apps/userreports/tests/test_pillow.py:396]:
```suggestion
            domain=cls.domain,
```

**Review by kaapstorm — APPROVED** [review]
This is looking good. Just a couple of suggestions.

File: corehq/apps/userreports/tests/test_pillow.py
Comment (SmittieC) [reply corehq/apps/userreports/tests/test_pillow.py]: https://github.com/dimagi/commcare-hq/pull/33781/commits/2d8752ed396d6ec49e6e3aeb8940ecc39f1ea970

File: corehq/apps/userreports/tests/test_pillow.py
Comment (SmittieC) [reply corehq/apps/userreports/tests/test_pillow.py]: https://github.com/dimagi/commcare-hq/pull/33781/commits/2d8752ed396d6ec49e6e3aeb8940ecc39f1ea970

File: corehq/apps/userreports/tests/test_pillow.py
Comment (mkangia) [inline corehq/apps/userreports/tests/test_pillow.py:463]: is this to help with the unexpected error in tests through github actions?
Is this still testing the same thing?

**Review by mkangia — CHANGES_REQUESTED** [review]
Hey @SmittieC 
Can you help understand the overall approach? I took a look at the spec as well.

I am not able to understand the need to create a new Log table for this. Can't we straight away create repeat records when a case/form is updated/submitted just like we do for usual case and form pillows, like [this](https://github.com/dimagi/commcare-hq/blob/5023fb2ce9d1b547fc4d95e861e91bd0690c590f/corehq/motech/repeaters/signals.py#L105) for instance?

**Comment by SmittieC** [comment]
> Hey @SmittieC Can you help understand the overall approach? I took a look at the spec as well.
> 
> I am not able to understand the need to create a new Log table for this. Can't we straight away create repeat records when a case/form is updated/submitted just like we do for usual case and form pillows, like [this](https://github.com/dimagi/commcare-hq/blob/5023fb2ce9d1b547fc4d95e861e91bd0690c590f/corehq/motech/repeaters/signals.py#L105) for instance?

@mkangia Sure! So one big difference I need to mention between this and current repeaters are that this needs to forward changes to a datasource, not raw forms or cases.

As for the apporach, imagine an audit log. It tells you 1. who did 2. what to 3. which tables. In this case we don't care about the "who", and only care about what changed on a datasource table. This change should be forwarded to the "subscriber" (or the remote connection), and they should be able to update their version of the datasource accordingly.

The reason for the log table is so the `RepeatRecord` can point to a specific record that holds "what changed" along with "how it changed". Current repeat records point to forms/cases, by now we need to point to a _change_. It's worth mentioning that multiple repeat records should be able to point to the same change - this way we support muliple "subscribers". We also get this for free by having this table. Do feel free to set up a call if you still have questions or want to clarify something :smiley_cat:

**Review by Charl1996 — APPROVED** [review]

File: corehq/apps/userreports/util.py
Comment (mkangia) [reply corehq/apps/userreports/util.py:365]: That makes sense.
I like it when the check for feature flag is exposed as early as possible.
So, if it makes sense, at least the feature flag check can be moved to places where the method is being called to just highlight it to someone, if they are reading code, earlier, and not when they visit this method.

File: corehq/motech/repeaters/signals.py
Comment (mkangia) [inline corehq/motech/repeaters/signals.py:122]: Why is this needed here? I thought we were making repeat records when updating data source?

File: corehq/apps/userreports/sql/adapter.py
Comment (mkangia) [inline corehq/apps/userreports/sql/adapter.py:180]: nit: a better name would be `register_data_source_row_change`

**Comment by mkangia** [comment]
Thanks for the response @SmittieC 
I was thinking that we could create the row on the fly using the data source configuration, instead of saving the row to the table, which would remove the need to have a separate log table and we could simply have the case/form is like other case/form repeaters.
Having the data to be forwarded on the table does reduce the computation HQ would have needed to do, to form the data to be forwarded, so I think its okay to keep a separate log then.

One follow up:
1. How are we removing logs? Can we delete the log after the record has been forwarded? Or that discussion is being taken elsewhere?

File: corehq/motech/repeaters/repeater_generators.py
Comment (mkangia) [inline corehq/motech/repeaters/repeater_generators.py:756]: nit: not sure how this is intended to be used but how about adding a "row_id" or "doc_id" outside the data for easy fetching of the ID/row that needs to be updated?

File: corehq/motech/repeaters/signals.py
Comment (SmittieC) [reply corehq/motech/repeaters/signals.py:122]: Yeah this is where we do it. We call this signal, which then creates the transaction log, and then calls the `create_repeat_record` method to create the record

File: corehq/motech/repeaters/repeater_generators.py
Comment (SmittieC) [reply corehq/motech/repeaters/repeater_generators.py]: Agreed. The `row_id` is the `doc_id`. See [this](https://github.com/dimagi/commcare-hq/blob/63a9f23802a6c6d9571e628684274c1bec201c97/corehq/motech/repeaters/signals.py#L112). The reason for naming it `row_id`, is basically because it is the `row_id` of the Datasource table

File: corehq/motech/repeaters/signals.py
Comment (mkangia) [reply corehq/motech/repeaters/signals.py:122]: and this when we save rows?
https://github.com/dimagi/commcare-hq/pull/33781/files#diff-1d71ce3e4eed9626ae52cd181a05d9a240094f731406d696625757de1d4efd97R147
Is that for something else?

Comment (SmittieC) [reply corehq/motech/repeaters/signals.py:122]: I'm not following. Mind rephrasing the question please?

Comment (mkangia) [reply corehq/motech/repeaters/signals.py:122]: Hey @SmittieC

My bad.
Confirming that this is just linking signal `ucr_data_source_updated` to the function `create_data_source_transaction_log_entry` and not actually calling it.
I misunderstood this as a call to the function `create_data_source_transaction_log_entry` which would create a log entry.

File: corehq/motech/repeaters/repeater_generators.py
Comment (mkangia) [reply corehq/motech/repeaters/repeater_generators.py]: Cool. I meant within the json here, there could be a row_id property that the receiver could use directly instead of finding that in the data property.

File: corehq/motech/repeaters/signals.py
Comment (SmittieC) [reply corehq/motech/repeaters/signals.py:122]: All good :+1:

**Comment by mkangia** [comment]
@SmittieC 
Can we cherry-pick relevant commits and start with a new PR instead for sake of simplicity and clarity?

**Comment by SmittieC** [comment]
Closing in favor of [this PR](https://github.com/dimagi/commcare-hq/pull/33877)

**Pull request closed without merging**
